### PR TITLE
fix: #3334 await realtime background tasks during cleanup

### LINKED ISSUE #3334: Realtime cleanup should await cancelled background tasks
### Please read this first

- **Have you read the docs?** Yes. This concerns Realtime session cleanup behavior rather than a documented usage option.
- **Have you searched for related issues?** Yes. I searched upstream issues and PRs. The closest prior work I found is #1976, which covered guardrail cleanup only and was closed without merge.

### Describe the bug

`RealtimeSession._cleanup()` cancels pending Realtime guardrail and tool-call background tasks, but the current cleanup helpers clear the tracking sets immediately after `task.cancel()`. They do not await the cancelled tasks with `asyncio.gather(..., return_exceptions=True)` or an equivalent exception-collecting wait.

Impact:

- `_cleanup()` can return before cancelled task `finally` blocks have run.
- Cancellation exceptions or task callbacks can be processed after the session has already moved on to model close.
- Guardrail and tool-call task cleanup behaves differently from the usual asyncio cancel-and-await cleanup pattern.

Relevant edge cases considered:

- Guardrail background tasks and tool-call background tasks should both be awaited during cleanup.
- Cancelled task `finally` blocks should complete before `_cleanup()` returns.
- Cancellation exceptions should be collected with `return_exceptions=True`.
- Tracking sets should be cleared only after the cancelled tasks have been awaited.

### Debug information

- Agents SDK version: upstream `main` at `479640e`
- Python version: Python 3.12.1

### Repro steps

Run this minimal script from a repository checkout.

```python
import asyncio
from typing import Any

from agents.realtime.agent import RealtimeAgent
from agents.realtime.model import RealtimeModel
from agents.realtime.session import RealtimeSession


class FakeRealtimeModel(RealtimeModel):
    def add_listener(self, listener: Any) -> None:
        pass

    def remove_listener(self, listener: Any) -> None:
        pass

    async def connect(self, options: Any) -> None:
        pass

    async def send_event(self, event: Any) -> None:
        pass

    async def close(self) -> None:
        pass


async def main() -> None:
    session = RealtimeSession(FakeRealtimeModel(), RealtimeAgent(name="agent"), None)
    guardrail_started = asyncio.Event()
    guardrail_finished = asyncio.Event()
    tool_started = asyncio.Event()
    tool_finished = asyncio.Event()

    async def guardrail_task() -> None:
        guardrail_started.set()
        try:
            await asyncio.Event().wait()
        finally:
            await asyncio.sleep(0)
            guardrail_finished.set()

    async def tool_call_task() -> None:
        tool_started.set()
        try:
            await asyncio.Event().wait()
        finally:
            await asyncio.sleep(0)
            tool_finished.set()

    guardrail = asyncio.create_task(guardrail_task())
    tool_call = asyncio.create_task(tool_call_task())
    session._guardrail_tasks.add(guardrail)
    session._tool_call_tasks.add(tool_call)

    await guardrail_started.wait()
    await tool_started.wait()

    await session._cleanup()

    print(f"guardrail_done_after_cleanup: {guardrail.done()}")
    print(f"tool_done_after_cleanup: {tool_call.done()}")
    print(f"guardrail_finally_ran: {guardrail_finished.is_set()}")
    print(f"tool_finally_ran: {tool_finished.is_set()}")
    print(f"tracked_guardrail_tasks: {len(session._guardrail_tasks)}")
    print(f"tracked_tool_call_tasks: {len(session._tool_call_tasks)}")


asyncio.run(main())
```

Actual result on upstream `main` at `479640e`:

```text
guardrail_done_after_cleanup: False
tool_done_after_cleanup: False
guardrail_finally_ran: False
tool_finally_ran: False
tracked_guardrail_tasks: 0
tracked_tool_call_tasks: 0
```

### Expected behavior

`RealtimeSession._cleanup()` should cancel pending guardrail and tool-call background tasks, await the tracked tasks while collecting cancellation exceptions, and only then clear the tracking sets and continue shutdown.

### PR DESCRIPTION
**Pull request by Aphroq**
### Summary

- Await cancelled Realtime guardrail and tool-call background tasks during session cleanup before clearing their tracking sets.
- Keep cleanup exception handling exercised with real asyncio tasks, and add a regression test that verifies cancelled tasks finish their `finally` blocks before `_cleanup()` returns.

Related: #1976  previously identified the missing await for Realtime guardrail task cleanup. This PR keeps that direction, also covers Realtime tool-call task cleanup, and adds regression coverage.

### Test plan

- `uv run pytest tests/realtime/test_session.py::test_cleanup_awaits_cancelled_background_tasks tests/realtime/test_session_exceptions.py::TestSessionExceptions::test_exception_during_guardrail_processing`
- `bash .agents/skills/code-change-verification/scripts/run.sh`

### Issue number

Closes #3334

### Checks

- [x] I've added new tests (if relevant)
- [ ] I've added/updated the relevant documentation
- [x] I've run `make lint` and `make format`
- [x] I've made sure tests pass

### CODE REVIEW

**Comment by Aphroq** [comment]
@codex  review

File: src/agents/realtime/session.py [hunk 1]
```diff
@@ -1060,11 +1066,17 @@ def _on_tool_call_task_done(self, task: asyncio.Task[Any]) -> None:
...
+        tasks = list(self._tool_call_tasks)
+        if not tasks:
+            return
+
+        for task in tasks:
             if not task.done():
                 task.cancel()
-        self._tool_call_tasks.clear()
+
+        await asyncio.gather(*tasks, return_exceptions=True)
```
[the remark this answers is not in the document]
Comment (Aphroq) [reply src/agents/realtime/session.py]: I changed Realtime background task cleanup to use a bounded wait: cleanup still cancels guardrail and tool-call background tasks and gives them a short grace period to run `finally` / cancellation cleanup, but if a task still does not exit within that window, we log a warning and continue session shutdown instead of letting `close()` / `__aexit__` hang indefinitely on user code.

This applies to both guardrail tasks and tool-call tasks, while preserving the existing behavior for already-finished tasks and normally cancellable tasks:

- Finished tasks are not cancelled again.
- Tasks that respond to cancellation are still awaited, so their `finally` blocks can finish before `_cleanup()` returns.
- Tasks that suppress or delay `CancelledError` no longer make cleanup wait forever.
- After the bounded wait, tracking sets are cleared so model close and event iterator wakeup can continue.

**Comment by Aphroq** [comment]
@codex  review

**Comment by Aphroq** [comment]
Updated the branch with one more shutdown hardening pass :

- Guardrail and tool-call background task cleanup now waits concurrently, so the worst case does not stack two cleanup grace periods sequentially.
- Added an internal `_closing` guard so late background task callbacks after a cleanup timeout do not write `_stored_exception` or enqueue `RealtimeError` events while the session is closing/closed.
- Added regression coverage for:
  - guardrail/tool-call cleanup waiting concurrently
  - background tasks that fail after the timeout not polluting a closed session

About the timeout: `_BACKGROUND_TASK_CLEANUP_TIMEOUT_SECONDS` remains a private best-effort cleanup grace period, not public configuration or a completion guarantee. It is only meant to give normally cancellable tasks a chance to run `finally` / cancellation cleanup while keeping `close()` / `__aexit__` bounded if user code suppresses or delays `CancelledError`. The previous behavior cancelled tasks and immediately cleared the tracking sets; this keeps shutdown bounded while giving cancellation cleanup a better chance to run.

**Comment by seratch** [comment]
Thanks for working through the shutdown edge cases. The bounded concurrent wait and late-callback tests are useful, but this head now conflicts with main and still fails when a tracked background task calls close() itself. It also does not prevent new tasks from being enqueued while cleanup is waiting.

I suggest moving the bounded-wait and late-callback coverage into #3656, which already handles the current-task and in-flight event races, and closing this PR as superseded.
